### DISCUSSION TITLE
Graphql query with Next Auth Session token

### DISCUSSION THREAD

**Discussion by wchorski** [keystonejs/keystone#8863] [Questions] [answered]
## Stack

my repo - [https://github.com/wchorski/KYPN/tree/integrated-ks-next](https://github.com/wchorski/KYPN/tree/integrated-ks-next)

I'm attempting to combined these two examples
- [framework-nextjs-app-directory](https://github.com/keystonejs/keystone/tree/main/examples/framework-nextjs-app-directory)
- [custom-session-next-auth](https://github.com/keystonejs/keystone/tree/main/examples/custom-session-next-auth)

attaching the session auth when querying / mutating from the client via `graphql-request` is where I'm struggling. (I'm using `graphql-yoga` as the server).

I did see this note in `/src/pages/graphql.ts`

```ts
/*
    `keystoneContext` object doesn't have user's session information.
    You need an authenticated context to CRUD data behind access control.
    keystoneContext.withRequest(req, res) automatically unwraps the session cookie
    in the request object and gives you a `context` object with session info
    and an elevated sudo context to bypass access control if needed (context.sudo()).
  */
```

but I'm assuming that's referring to **keystone's** built in auth library.

## Attempt 1
I took some notes from [this repo](https://github.com/egilkenson/keystone-next-auth) to see if i can send the session token via `Authorization: Bearer TOKEN` header.

helper function
```ts
export async function fetchProtectedRoute(req:NextRequest, query:string, variables = {}): Promise<unknown> {


  const cookieStore = cookies()
  const token = cookieStore.get('next-auth.session-token')
  console.log({token});

  const client = new GraphQLClient(envs.FRONTEND_URL + '/api/graphql')
  const requestHeaders = {
    ...req.headers,
    authorization: `Bearer ${token?.value}`
  }
  return client.request(query, variables, requestHeaders)
}
```

fetching from server component
```ts
const data = await fetchProtectedRoute(req, query, variables)
console.log({data});
```

when called the query is successful, but returns an empty array (meaning the query wasn't authenticated)

## Attempt 2
Maybe I need to pass **Next-Auth's** session into the created Yoga API but no dice
```ts
context: async ({ req, res }) => {

    const session = await getServerSession(nextAuthOptions)

    return keystoneContext.withSession(session).withRequest(req, res)
  },

// error - "Invariant: headers() expects to have requestAsyncStorage, none available."
```

 
## Attempt 3
With mutations I'm even more lost. I understand this is an illegal move but posting here to get the idea of what I'm trying to accomplish.

```ts
const response = await keystoneContext.withSession(session).mutation.addToCart({
    addToCartId,
    productId,
  })
```

## Bonus

bonus points for how to convert the `/pages/api/graphql.ts` to '/app/api/graphql/route.ts`. 

`/pages/api/graphql.ts`
```ts
import { createYoga } from 'graphql-yoga';
import type { NextApiRequest, NextApiResponse } from 'next';
import { keystoneContext } from '@ks/context';

export const config = {
  api: {
    // Disable body parsing (required for file uploads)
    bodyParser: false,
  },
};

// Use Keystone API to create GraphQL handler
export default createYoga<{
  req: NextApiRequest;
  res: NextApiResponse;
}>({
  graphqlEndpoint: '/api/graphql',
  schema: keystoneContext.graphql.schema,

  context: ({ req, res }) => keystoneContext.withRequest(req, res),
});
```

I tried to convert it with this code (without typecript whining) but got a `500` error
```ts
import { createYoga } from 'graphql-yoga';
import type { NextApiRequest, NextApiResponse } from 'next';
import { keystoneContext } from '@ks/context';

export const config = {
  api: {
    // Disable body parsing (required for file uploads)
    bodyParser: false,
  },
};

const { handleRequest } = createYoga<{
    req: NextApiRequest;
    res: NextApiResponse;
  }>({
  graphqlEndpoint: '/api/graphql',
  schema: keystoneContext.graphql.schema,

  context: ({ req, res }) => keystoneContext.withRequest(req, res),
})

export { handleRequest as GET, handleRequest as POST }
```

This would be a huge hurdle to overcome, and I'd be happy to display this as one of my favorite Open Source full stack frameworks.

**Comment by marekryb** [comment, marked as answer]
Hi,

I have this working with standard setup like this:
```typescript
// pages/api/graphql.ts
const keystoneContext = await getKeystoneContext();
return createYoga({
      ...
      context: ({ req, res }) => keystoneContext.withRequest(req, res),
      ...
})(req,res)
```
`withRequest` is already using `withSession`. Not sure where your `headers() expects...` error comes from, but I am guessing from `getServerSession`. I am pretty sure you can just remove it. https://github.com/keystonejs/keystone/blob/41276b6877c0c1e9332d8c3e37ccd772a5b2c48b/packages/core/src/lib/context/createContext.ts#L83

You can use mutation if you go through graphql layer api:
```typescript
const yourAsyncFunction = async () => {
  const session = await getServerSession(nextAuthOptions);
  const keystoneContext = await getKeystoneContext();
  const dataOrResult = await keystoneContext.withSession(session).graphql.run({
    // This is plain graphql query as in playground. Can be string or DocumentNode (eg. from graphql-codegen)
    query: yourQueryOrMutation, 
    variables: your variables,
  });
  ...
}

```
Cannot help you with `app/api/graphql` thingy, sorry. Tried it myself, didn't work (different error), added to todo list with lowest priority and moved on ;-)

**Reply by marekryb** [reply]
Regarding `headers() expects` - I checked and one subtle difference to the keystone example is that I use `getToken` instead of `getServerSession` inside SessionStrategy getter. 
```typescript
const token = await getToken({ req: { headers, cookies } as any });
```
If the error does not go away, you might give it a try.

**Comment by wchorski** [comment]
@marekryb Thanks for filling the gap in my context knowledge.

Also, I did forget to wrap the app in the Next-Auth Session Provider so that the cookie is sent on client requests

`providers.tsx`
```ts
'use client'
// cred Ethan - https://www.youtube.com/watch?v=2kgqPvs0j_I

import { SessionProvider } from 'next-auth/react'

type Props = {
  children?: React.ReactNode
}

export const Providers = ({ children }: Props) => {
  return <SessionProvider>{children}</SessionProvider>
}
```

**Comment by borisno2** [comment]
Hey @wchorski 
For the bonus points I have `/api/graphql` working in the app directory in [NextJS in my app](https://github.com/borisno2/on-the-hill-drama-club/blob/main/src/app/api/graphql/route.ts) I did have the same error you were seeing in `Attempt 2` so had to wrap the next-auth `getServerSession` in a special function. I am using Apollo, but it should work in a similar way with Yoga. I am using `context.withSession` and passing in the next-auth session directly as `context.withRequest` does not really align with the NextJS app router.
